### PR TITLE
[dagster-airlift][rfc] support time-window partitioned assets

### PR DESCRIPTION
**Pull request by dpeng817**
## Summary & Motivation
Support automatically munging time window partitioned assets into partitioned materializations via the sensor.

I've placed the main implementation in its own function, the signature of which I'm thinking is what we'll be exposing to users as a pluggability point.
## How I Tested These Changes
Added a bunch of new unit tests.

## Changelog
NOCHANGELOG

### CODE REVIEW

**Comment by dpeng817** [comment]
* **#24773** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24773?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @dpeng817 and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: examples/experimental/dagster-airlift/dagster_airlift/core/partition_utils.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    )
+    check.invariant(
+        logical_date.timestamp() >= partitions_def.start.timestamp(),
+        (
+            f"For run {task_instance.run_id} in dag {task_instance.dag_id}, "
+            f"Logical date is before the start of the partitions definition in asset key {asset_key.to_user_string()}. "
+            "Ensure that the start date of your PartitionsDefinition is early enough to capture current runs coming from airflow."
+        ),
+    )
+    return partitions_def.get_partition_key_for_timestamp(logical_date.timestamp())
```
Comment (schrockn) [inline examples/experimental/dagster-airlift/dagster_airlift/core/partition_utils.py:53]: we are definitely going to need to make this pluggable at some level

**Review by schrockn — DISMISSED** [review]
Makes sense overall. I think it is going to be important to at least hypothesize what the user-pluggability story is going to be as users will want complete control over the format of the partition keys in the general case.

File: examples/experimental/dagster-airlift/dagster_airlift/core/partition_utils.py (hunk 1 above)
Comment (schrockn) [reply examples/experimental/dagster-airlift/dagster_airlift/core/partition_utils.py:53]: Yeah would like to see that downstack PR before stamping this

**Comment by dpeng817** [comment]
see https://github.com/dagster-io/dagster/pull/24807 @schrockn

File: examples/experimental/dagster-airlift/dagster_airlift/core/partition_utils.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+def get_partition_key_from_task_instance(
+    partitions_def: PartitionsDefinition, task_instance: TaskInstance, asset_key: AssetKey
+) -> str:
+    partitions_def = check.inst(
+        partitions_def,
+        TimeWindowPartitionsDefinition,
+        "Only time window-partitioned assets are supported.",
+    )
+    logical_date = task_instance.logical_datetime
```
Comment (schrockn) [inline examples/experimental/dagster-airlift/dagster_airlift/core/partition_utils.py:20]: when do we use the name "logical" versus "effective"?

File: examples/experimental/dagster-airlift/dagster_airlift/core/airflow_instance.py [hunk 3]
```diff
@@ -385,6 +377,15 @@ def state(self) -> str:
...
         return self.metadata.get("note") or ""
 
+    @property
+    def logical_datetime(self) -> datetime.datetime:
+        # In airflow < 2.2, execution_date is set instead of logical_date.
+        logical_date_str = check.not_none(
+            self.metadata.get("logical_date") or self.metadata.get("execution_date"),
+            "Expected one of execution_date or logical_date to be set.",
+        )
+        return datetime.datetime.fromisoformat(logical_date_str)
```
Comment (schrockn) [inline examples/experimental/dagster-airlift/dagster_airlift/core/airflow_instance.py:387]: oh I see this is airflow terminology

File: examples/experimental/dagster-airlift/dagster_airlift/core/partition_utils.py (hunk 2 above)
Comment (schrockn) [inline examples/experimental/dagster-airlift/dagster_airlift/core/partition_utils.py:20]: To make this super clear, I think we should prefix local variables that store airflow's concept of date with `airflow_`. It is very easy IMO to get all the different notions of time confused in this function. Let's do a readability pass.

**Review by schrockn — CHANGES_REQUESTED** [review]
Let's do a pass to make sure we don't confuse the different notions of time floating around here.

**Pull request closed without merging**